spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameterManager.scala

```diff
@@ -0,0 +1,333 @@
+ * here serves as a parameter server.
+ *
+ * A Tensor is sliced into `executorNum` chunks and each chunk is assigned to a particular node
+ * (Spark executor). Likewise, gradients for each chunk are also assigned and stored on separate
+ * nodes. In this way, gradient aggregation and parameter updates can be performed independently for
+ * each chunk on separate nodes.
+ *
+ * @param id distinguish from other parameters
+ * @param executorId executorId of the node
+ * @param executorNum how many executors will use this parameter
+ * @param partitionNum partition numbers
+ * @param size size of the parameter (1D vector)
+ * @param master ParameterManagerMaster endpoint
+ */
+class AllReduceParameterManager(val id: Int, val executorId: Int, executorNum: Int,
+  partitionNum: Int, size: Int, val master: ParameterManagerMaster) {
+  import AllReduceParameterManager._
+
+  private val syncPoolSize: Int = System.getProperty(
+    "bigdl.Parameter.syncPoolSize", "4").toInt
+
+  val syncPool = Executors.newFixedThreadPool(syncPoolSize, new ThreadFactory {
+    override def newThread(r: Runnable): Thread = {
+      val t = Executors.defaultThreadFactory().newThread(r)
+      t.setDaemon(true)
+      t
+    }
+  })
+
+  var job1Start: Boolean = false
+  var initFinished = false
+
+  val taskSize = size / executorNum
+  val extraSize = size % executorNum
+
+  /**
+    * This method should be called on each RDD partition before parameter synchronization begins.
+    * An empty gradient tensor is placed in the block manager that can be used to store gradients.
+    * A 1 / executorNum fraction of the `parameter` tensor is copied to the block manager as a
+    * compressed tensor.
+    *
+    * @param parameter A tensor representing the initial underlying weights of this
+    *                  `AllReduceParameterManager`
+    */
+  def init[T: ClassTag](parameter: Tensor[T], state: Table)
+    (implicit ev: TensorNumeric[T]): Unit = {
+    val _classTag = classTag[T]
+    val start = executorId * taskSize + math.min(executorId, extraSize)
+    val length = taskSize + (if (executorId < extraSize) 1 else 0)
+    val _weightsExecutor = Tensor[T](length)(_classTag, ev).copy(parameter.narrow(1,
+      start + 1, length))
+    BlockManagerWrapper.putSingle(getWeightExecutorId(),
+      _weightsExecutor, StorageLevel.MEMORY_AND_DISK, tellMaster = false)
+
+    BlockManagerWrapper.putSingle(getWeightId(),
+      parameter, StorageLevel.MEMORY_AND_DISK, tellMaster = false)
+
+    val _gradientsExecutor = Tensor[T](length)(_classTag, ev)
+    BlockManagerWrapper.putSingle(getGradientExecutorId(),
+      _gradientsExecutor, StorageLevel.MEMORY_AND_DISK, tellMaster = false)
+
+    val blockId = getWeightBlockId(executorId)
+    val fp16param = new FP16CompressedTensor[T](length)(_classTag)
+    fp16param.compress(0, parameter, start, length)
+    BlockManagerWrapper.putBytes(blockId, fp16param.bytes(), StorageLevel.MEMORY_ONLY_SER)
```

Comment:
what if two partitions are scheduled to the same executor? will this init method be executed twice? Is it a problem?